src/python/WMCore/BossAir/Plugins/BasePlugin.py

```diff
@@ -181,3 +181,31 @@ def scramArchtoRequiredArch(scramArch=None):
             archs = defaultArch
 
         return archs
+
+    @staticmethod
+    def cudaCapabilityToSingleVersion(capabilities=None):
+        """
+        Given a list of CUDA capabilities (with strings in a version style),
+        finds the smallest version required and convert it to a single integer
+        for comparison/job matchmaking purposes.
+        Version conversion formula is: (1000 * major + 10 * medium + minor)
+        :param capabilities: a list of string versions
+        :return: an integer with the version value; 0 in case of failure
```

Comment:
What is the order here ... is `0` with least significance or vise versa?

Reply:
I expect the job matchmaking to be `available_version >= required_version`, so 0 here would mean that any cuda capability is supported by the job. I am not sure this answers your question, otherwise I might have not understood your question.

Reply:
Partially it does.
What bothers me here, is that default value is again `0` and it overlaps with the value returned in case of an error/failure. Which would be a valid case if 0 was never to be reached in the `matchmaking` condition. But since it is included, then it would trigger similar behavior in failure and by default.

Reply:
Yes, it can return from a version to integer conversion or from an error. However, this code will only be executed if GPUs are required, and if they are, then there are validated GPU parameters and the version is enforced to be something like `r"^\d+\.\d+(\.\d+)?$"`. IF a version provided is 0.0, then yes, we get a "valid" 0 single version, which we could get if an empty list had been provided. But then again, everything is validated at the spec level.

I could change it to None and then mark the classad as undefined. But then it will be a job requiring GPU but asking for an undefined CUDA capability. I am not sure whether that could trigger other errors and unwanted behavior or not. Please let me know if you have any suggestions.

Reply:
Alan, I do not want to say it... but usually there are many `if`s in the justification process of an unhealthy code behavior. Overlapping of  `default` and `failure` return values is not a good practice in general. Let alone type modification.  I know we are supposed to be protected on another level, and we should not enter this bit of code under bad conditions, but lets imagine something changes in the future and we change our policy regarding GPU job execution as well.... Then we may enter this code under expected conditions, or because of a bug in the request validation code.... anything may happen. 

So I'd bet for returning a `None` value in case of a failure and let the upstream code deal with it. But that's up to you. If you think we are 100% safe and we are about to call this function under the best condition only at any possible moment in the future, we can leave it as it is.

Reply:
I don't really have a strong preference here. I can make it return None by default, have a check in SingleCondorPlugin for None and map that to `undefined` classad. Sounds good?

Reply:
sounds good, yes
thank you @amaltaro